File: code/SoundPlayground/SoundPlayground/Core/Note.cs
using System.Collections.Generic;

namespace SoundPlayground.Core {
    public enum NoteAccidental {
        Flat = 0, None = 1, Sharp = 2
    }

    public class Note {
        protected static Dictionary<char, int> PitchClassesDictionary = new Dictionary<char, int>() {
            [ 'C' ] = 0,
            [ 'D' ] = 2,
            [ 'E' ] = 4,
            [ 'F' ] = 5,
            [ 'G' ] = 7,
            [ 'A' ] = 9,
            [ 'B' ] = 11
        };

        public static int ParsePitchClass ( string c ) {
            return ParsePitchClass( c.ToCharArray()[ 0 ] );
        }

        public static int ParsePitchClass ( char c ) {
            return PitchClassesDictionary[ char.ToUpper( c ) ];
        }

        public int Start { get; set; } = 0;

        public int PitchClass { get; set; }
        
        public int Duration { get; set; } = 4;

        public int Octave { get; set; } = 4;
        
        public NoteAccidental Accidental { get; set; } = NoteAccidental.None;

        public int ToInteger () {
            int accidental = ( ( int )Accidental - 1 );

            return Octave * 12 + PitchClass + accidental;
        }

        // public static NoteNode FromInteger ( int node ) {

        // }

        public static implicit operator int ( Note note ) => note.ToInteger();

        // public static implicit operator NoteNode ( int note ) => NoteNode.FromInteger( note );
    }
}

File: code/SoundPlaygroundPy/core/note.py
class MusicEvent():
    def __init__ ( self, timestamp = 0 ):
        self.timestamp = timestamp

class NoteAccidental():
    DOUBLEFLAT = 0
    FLAT = 1
    NONE = 2
    SHARP = 3
    DOUBLESHARP = 4

class Note( MusicEvent ):
    def __init__ ( self, timestamp = 0, pitch_class = 0, duration = 4, octave = 4, channel = 0, velocity = 127, accidental = NoteAccidental.NONE ):
        super().__init__( timestamp )

        self.pitch_class = pitch_class
        self.duration = duration
        self.octave = octave
        self.channel = channel
        self.velocity = velocity
        self.accidental = accidental

    def __int__ ( self ):
        accidental = self.accidental - 2

        return ( self.octave + 1 ) * 12 + self.pitch_class + accidental

    def __str__ ( self ):
        return f"<Note Timestamp={self.timestamp} PitchClass={self.pitch_class} Duration={self.duration} Octave={self.octave} Accidental={self.accidental} Channel={self.channel} Velocity={self.velocity}>"
